cclib/method/ddec.py

```diff
@@ -271,59 +281,105 @@ def calculate(self, indices=None, fupdate=0.05):
             self.radial_grid_r.append(r)
 
         # Carry out step 3 of DDEC6 algorithm [Determine conditioned charge density and tau]
-        self.logger.info("Conditioning charge densities.")
+        self.logger.info("Conditioning charge densities. (Step 3/7)")
         self.condition_densities()
 
-        # STEP 4-7
-        # Optimize weights on each grid point
-        kappa = [0.0 for x in self.data.atomnos]
+        # Steps 4 through 7 contain similar routines. Comments the precede each step explain the
+        # differences among them.
+
+        # * STEP 4 *
+        # In step 4, calculate w, u, g, and h but always skip kappa updates.
+        self._kappa = [0.0] * self.data.atomnos
         update_kappa = False
+        # N_A is assigned number of electrons determined using equation 72.
         self.N_A = []
-        steps = 4
-        while steps < 5:  # change this to 8 after implementing kappa routine
-            self.logger.info("Optimizing charge weights (Step {}).".format(steps))
+        # u_A is determined using equation 77.
+        self.u_A = []
+
+        self.logger.info("Optimizing grid weights. (Step 4/7)")
+        self.N_A.append(self._calculate_w_and_u())
+
+        # Calculate G_A and H_A to update w_A based on S4.3 in doi: 10.1039/c6ra04656h
+        self.reshape_G()
+        self.calculate_H()
+
+        # Update weights (w_A) using equation 96 in doi: 10.1039/c6ra04656h
+        for atomi in range(len(self._cond_density)):
```

Comment:
This is the first time `self._cond_density` shows up  when reading through `calculate`. I'm sure it's set somewhere in something called above, but that's not obvious. You're also overwriting it... so why not just create from scratch?

Also what is `range(len(self._cond_density))` really iterating over? If I'm looking at the right loop below, it's simply iterating over `natom`, so why not use that here?

Reply:
`self._cond_density` is first populated in Step 3, within the function `self.condition_densities()`. I've moved the declaration of the list to the calculate() function to prevent confusion. In each steps that follow (4-6), `self._cond_density` is updated with new values.

I've changed the range to use natom instead. Thanks.